src/Test/TestAIMuriel.java

```diff
@@ -55,37 +60,65 @@ private void cleanupTest()
   public boolean runTest()
   {
     boolean testPassed = true;
-    testPassed &= validate(testBuildMegatank(), "  Build Megatank test failed.");
-    testPassed &= validate(testHuntStall(), "  Hunting stall-test failed.");
-    testPassed &= validate(testClearAttackRoute(), "  Route clearing test failed.");
-    testPassed &= validate(testWalkInLine(), "  Line walking test failed.");
-    testPassed &= validate(testInfWadeThroughTanks(), "  Infantry move priority test failed.");
-    testPassed &= validate(testTankWadeThroughInfs(), "  Tank move priority test failed.");
+    testPassed &= validate(testForAllAIs((ai) -> testBuildMegatank(ai)), "  Build Megatank test failed.");
+    testPassed &= validate(testForAllAIs((ai) -> testHuntStall(ai)), "  Hunting stall-test failed.");
+    testPassed &= validate(testForAllAIs((ai) -> testClearAttackRoute(ai)), "  Route clearing test failed.");
+    testPassed &= validate(testForAllAIs((ai) -> testWalkInLine(ai)), "  Line walking test failed.");
+    testPassed &= validate(testForAllAIs((ai) -> testInfWadeThroughTanks(ai)), "  Infantry move priority test failed.");
+    testPassed &= validate(testTankWadeThroughInfs(Muriel.info), "  Tank move priority test failed.");
+    // TODO: Consider making Wally pass the above
+//    testPassed &= validate(testForAllAIs((ai) -> testProductionClearing(ai)), "  Infantry move priority test failed.");
+    testPassed &= validate(testProductionClearing(WallyAI.info), "  Free up industry test failed.");
     return testPassed;
   }
 
-  /** Confirm that Muriel will build the correct counter, even when there is only one possible counter. */
-  private boolean testBuildMegatank()
+  private boolean testForAllAIs(Function<AIMaker, Boolean> test)
   {
-    setupTest();
+    boolean testPassed = true;
+
+    for( AIMaker ai : ais )
+    {
+      testPassed &= test.apply(ai);
+      if( !testPassed )
+        break;
+    }
+
+    return testPassed;
+  }
+
+  /** Confirm that the AI will build the correct counter, even when there is only one possible counter. */
+  private boolean testBuildMegatank(AIMaker ai)
+  {
+    setupTest(ai);
 
     // Add an enemy to counter.
     addUnit(testMap, testCo2, "Megatank", 1, 1);
 
-    // Put Muriel in control
-    testCo1.setAIController(testAI);
-
     // Give Muriel resources.
     testCo1.money = 80000;
 
     // Ask Muriel what to do.
     testCo1.initTurn(testMap);
-    GameAction act = testCo1.getNextAIAction(testMap);
-    performGameAction(act, testGame);
+    GameAction act = null;
+    boolean testPassed = true;
+    do
+    {
+      act = testCo1.getNextAIAction(testMap);
+      if( null != act )
+        testPassed &= validate(performGameAction(act, testGame), "    "+ai.getName()+" generated a bad action!");
```

Comment:
if `null == act` seems like it should also be a test failure.
Also why turn this into a loop? There's like, nothing that can be done except unit production anyway. One action should be all that is required.

Reply:
Unless Wally is simply building other things first instead - which may be forgivable. This might be a case for 1) reducing the available funds to force the issue, and b) this being more of a Muriel-specific tests - who's to say that Artillery isn't the correct response, generally speaking? If we are creating an AI obstacle course, this test probably doesn't belong there.

Reply:
Yeah. Wally built an inf first, if I recall correctly.
Artillery doesn't answer a Megatank. It just doesn't do enough damage, and is too vulnerable (and has too little threat area).
Bombers could work, but there's no airport.

If you did want to force the issue, I'd suggest providing only one factory.

Reply:
If we really want to test "Building Megatank as the correct counter" then I would either reduce budget or factory count to ensure the AI performs correctly under the most constrained conditions.

Reply:
Eh... fiiiiiine